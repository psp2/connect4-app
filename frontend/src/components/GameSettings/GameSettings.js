import React, { useState } from "react";
import { FormControl } from "react-bootstrap";
import "../../pages/Menus/Menus.scss";
import "./GameSettings.scss";

function GameSettings(props) {
  const [options] = useState([1, 2, 3, 4, 5, 6, 7, 8, 9]);
  const [difficulty, setDifficulty] = useState(null);
  const [boardSize, setBoardSize] = useState(null);
  const [onlineMode, setOnlineMode] = useState(false);
  const [name1, setName1] = useState(null);
  const [name2, setName2] = useState(null);
  const [gameCode, setGameCode] = useState(null);

  function resetSettings() {
    props.resetPlayers(0);
    props.resetGameMode(0);
    props.resetSettings();
  }

  function startGame() {
    if (boardSize && (difficulty || name1)) {
      // If not in online mode, two names must be entered
      if (props.numPlayers === 2 && !onlineMode && !name2) {
        return;
      }
      props.submitSettings(difficulty, name1, name2, boardSize, gameCode);
    }
  }

  function selectDifficulty(diff) {
    setDifficulty(diff);
  }

  function selectName1(name) {
    setName1(name);
  }

  function selectName2(name) {
    setName2(name);
  }

  function selectSize(size) {
    setBoardSize(size);
  }

  function selectGameCode(code) {
    setGameCode(code);
  }

  function createDifficultyIcons(value) {
    if (value === difficulty) {
      return (
        <div className="SelectedValue" onClick={() => selectDifficulty(value)}>
          {value}
        </div>
      );
    } else {
      return (
        <div
          className="UnselectedValue"
          onClick={() => selectDifficulty(value)}
        >
          {value}
        </div>
      );
    }
  }

  function createSizeIcons(value) {
    if (value === boardSize) {
      return (
        <div className="SelectedValue" onClick={() => selectSize(value)}>
          {value}
        </div>
      );
    } else {
      return (
        <div className="UnselectedValue" onClick={() => selectSize(value)}>
          {value}
        </div>
      );
    }
  }

  return (
    <div className="MenuBoard">
      <div className="MenuTitle">
        <div className="MenuTitleWhite">Game</div>
        <div className="MenuTitleRed">Settings</div>
      </div>
      <div className="SettingsRow">
        {props.numPlayers === 1 && (
          <div className="DifficultySelect">
            <div className="SettingsHeading">
              <div className="ButtonCaption">Difficulty</div>
              <hr className="SettingsLine"></hr>
            </div>
            <div className="SettingsValueRow">
              {options.slice(0, 5).map((value) => createDifficultyIcons(value))}
            </div>
            <div className="SettingsValueRow">
              {options.slice(5, 9).map((value) => createDifficultyIcons(value))}
            </div>
          </div>
        )}
        {props.numPlayers !== 1 && (
          <div className="NameSelect">
            <div className="SettingsHeading">
              <div className="ButtonCaption">Enter Names</div>
              <hr className="SettingsLine"></hr>
            </div>
            {!onlineMode && (
              <div
                className="goOnlineButton"
                onClick={() => setOnlineMode(true)}
              >
                <div className="ButtonCaption">Switch to Online Mode</div>
              </div>
            )}
            {onlineMode && (
              <div
                className="goOfflineButton"
                onClick={() => setOnlineMode(false)}
              >
                <div className="ButtonCaption">Switch to Offline Mode</div>
              </div>
            )}
            <div className="nameZone">
              <FormControl
                className="nameInput"
                placeholder={"Player 1 Name"}
                defaultValue={""}
                onChange={(e) => selectName1(e.target.value)}
              />
              {!onlineMode && (
                <FormControl
                  className="nameInput"
                  placeholder={"Player 2 Name"}
                  defaultValue={""}
                  onChange={(e) => selectName2(e.target.value)}
                />
              )}
              {onlineMode && (
                <FormControl
                  className="nameInput"
                  placeholder={"Game Code"}
                  defaultValue={""}
                  onChange={(e) => selectGameCode(e.target.value)}
                />
              )}
            </div>
          </div>
        )}
        <div className="SizeSelect">
          <div className="SettingsHeading">
            <div className="ButtonCaption">Board Size</div>
            <hr className="SettingsLine"></hr>
          </div>
          <div className="SettingsValueRow">
            {options.slice(5, 8).map((value) => createSizeIcons(value))}
          </div>
        </div>
      </div>
      <div className="SettingsButtonRow">
        <div className="BackButtonYellow" onClick={resetSettings}>
          <div className="ButtonCaption">Back</div>
        </div>
        <div className="BackButtonGreen" onClick={startGame}>
          <div className="ButtonCaption">Start</div>
        </div>
      </div>
    </div>
  );
}

export default GameSettings;
