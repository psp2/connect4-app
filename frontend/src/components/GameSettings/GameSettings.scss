.SettingsRow {
  display: flex;
  flex-direction: row;
  margin-top: 3vw;
}

.SettingsHeading {
  margin-top: 2vw;
}

.SettingsLine {
  border: 2px solid black;
  width: 15vw;
}

.SettingsButtonRow {
  display: flex;
  flex-direction: row;
  margin-top: 4vw;
}

.DifficultySelect {
  border-radius: 50%;
  height: 20vw;
  width: 20vw;
  background-color: white;
  margin: auto;
  font-family: "Roboto", sans-serif;
  font-size: 2em;
  display: flex;
  flex-direction: column;
}

.SettingsValueRow {
  display: flex;
  flex-direction: row;
  margin: auto;
  margin-top: 1vh;
}

.UnselectedValue {
  width: 2vw;
  height: 2vw;
  border-style: solid;
  border-width: 1px;
  margin: auto;
}

.UnselectedValue:hover {
  background-color: rgb(15, 196, 15);
  transition-duration: 0.3s;
}

.SelectedValue {
  width: 2vw;
  height: 2vw;
  border-style: solid;
  border-width: 1px;
  margin: auto;
  background-color: rgb(15, 196, 15);
}

.NameSelect {
  border-radius: 50%;
  height: 20vw;
  width: 20vw;
  background-color: white;
  margin: auto;
  font-family: "Roboto", sans-serif;
  font-size: 2em;
  display: flex;
  flex-direction: column;
}

.SizeSelect {
  border-radius: 50%;
  height: 20vw;
  width: 20vw;
  background-color: red;
  margin: auto;
  font-family: "Roboto", sans-serif;
  font-size: 2em;
  display: flex;
  flex-direction: column;
}

.goOnlineButton {
  height: 2vw;
  width: 12vw;
  font-family: "Roboto", sans-serif;
  font-size: 0.5em;
  border-radius: 30px;
  color: white;
  background-color: #106eeb;
  margin: auto;
  margin-top: 1vh;
  margin-bottom: 3vh;
  display: flex;
  flex-direction: column;
}

.goOfflineButton {
  height: 2vw;
  width: 12vw;
  font-family: "Roboto", sans-serif;
  font-size: 0.5em;
  border-radius: 30px;
  color: white;
  background-color: red;
  margin: auto;
  margin-top: 1vh;
  margin-bottom: 3vh;
  display: flex;
  flex-direction: column;
}

.nameZone {
  margin: auto;
  margin-top: -1vh;
}

.nameInput {
  width: 12vw;
  height: 5vh;
  margin: auto;
  margin-bottom: 1vh;
  border-radius: 20px;
}

.nameInput:focus {
  outline-width: 0;
}
