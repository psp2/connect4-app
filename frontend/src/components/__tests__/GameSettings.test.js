import React from "react";
import GameSettings from "../GameSettings/GameSettings";
import renderer from "react-test-renderer";

it("GameSettings, renders correctly", () => {
  const tree = renderer.create(<GameSettings />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("GameSettings, Single Player", () => {
  // expect option for selecting difficulty to appear
  const tree = renderer.create(<GameSettings numPlayers={1} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("GameSettings, Multiplayer", () => {
  // expect option for entering names to appear
  const tree = renderer.create(<GameSettings numPlayers={2} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("GameSettings, null", () => {
  const tree = renderer.create(<GameSettings numPlayers={null} />).toJSON();
  expect(tree).toMatchSnapshot();
});
