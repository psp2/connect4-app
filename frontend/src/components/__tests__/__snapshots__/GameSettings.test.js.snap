// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameSettings, Multiplayer 1`] = `
<div
  className="MenuBoard"
>
  <div
    className="MenuTitle"
  >
    <div
      className="MenuTitleWhite"
    >
      Game
    </div>
    <div
      className="MenuTitleRed"
    >
      Settings
    </div>
  </div>
  <div
    className="SettingsRow"
  >
    <div
      className="NameSelect"
    >
      <div
        className="SettingsHeading"
      >
        <div
          className="ButtonCaption"
        >
          Enter Names
        </div>
        <hr
          className="SettingsLine"
        />
      </div>
      <div
        className="goOnlineButton"
        onClick={[Function]}
      >
        <div
          className="ButtonCaption"
        >
          Switch to Online Mode
        </div>
      </div>
      <div
        className="nameZone"
      >
        <input
          className="nameInput form-control"
          defaultValue=""
          onChange={[Function]}
          placeholder="Player 1 Name"
        />
        <input
          className="nameInput form-control"
          defaultValue=""
          onChange={[Function]}
          placeholder="Player 2 Name"
        />
      </div>
    </div>
    <div
      className="SizeSelect"
    >
      <div
        className="SettingsHeading"
      >
        <div
          className="ButtonCaption"
        >
          Board Size
        </div>
        <hr
          className="SettingsLine"
        />
      </div>
      <div
        className="SettingsValueRow"
      >
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          6
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          7
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          8
        </div>
      </div>
    </div>
  </div>
  <div
    className="SettingsButtonRow"
  >
    <div
      className="BackButtonYellow"
      onClick={[Function]}
    >
      <div
        className="ButtonCaption"
      >
        Back
      </div>
    </div>
    <div
      className="BackButtonGreen"
      onClick={[Function]}
    >
      <div
        className="ButtonCaption"
      >
        Start
      </div>
    </div>
  </div>
</div>
`;

exports[`GameSettings, Single Player 1`] = `
<div
  className="MenuBoard"
>
  <div
    className="MenuTitle"
  >
    <div
      className="MenuTitleWhite"
    >
      Game
    </div>
    <div
      className="MenuTitleRed"
    >
      Settings
    </div>
  </div>
  <div
    className="SettingsRow"
  >
    <div
      className="DifficultySelect"
    >
      <div
        className="SettingsHeading"
      >
        <div
          className="ButtonCaption"
        >
          Difficulty
        </div>
        <hr
          className="SettingsLine"
        />
      </div>
      <div
        className="SettingsValueRow"
      >
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          1
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          2
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          3
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          4
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          5
        </div>
      </div>
      <div
        className="SettingsValueRow"
      >
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          6
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          7
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          8
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          9
        </div>
      </div>
    </div>
    <div
      className="SizeSelect"
    >
      <div
        className="SettingsHeading"
      >
        <div
          className="ButtonCaption"
        >
          Board Size
        </div>
        <hr
          className="SettingsLine"
        />
      </div>
      <div
        className="SettingsValueRow"
      >
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          6
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          7
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          8
        </div>
      </div>
    </div>
  </div>
  <div
    className="SettingsButtonRow"
  >
    <div
      className="BackButtonYellow"
      onClick={[Function]}
    >
      <div
        className="ButtonCaption"
      >
        Back
      </div>
    </div>
    <div
      className="BackButtonGreen"
      onClick={[Function]}
    >
      <div
        className="ButtonCaption"
      >
        Start
      </div>
    </div>
  </div>
</div>
`;

exports[`GameSettings, null 1`] = `
<div
  className="MenuBoard"
>
  <div
    className="MenuTitle"
  >
    <div
      className="MenuTitleWhite"
    >
      Game
    </div>
    <div
      className="MenuTitleRed"
    >
      Settings
    </div>
  </div>
  <div
    className="SettingsRow"
  >
    <div
      className="NameSelect"
    >
      <div
        className="SettingsHeading"
      >
        <div
          className="ButtonCaption"
        >
          Enter Names
        </div>
        <hr
          className="SettingsLine"
        />
      </div>
      <div
        className="goOnlineButton"
        onClick={[Function]}
      >
        <div
          className="ButtonCaption"
        >
          Switch to Online Mode
        </div>
      </div>
      <div
        className="nameZone"
      >
        <input
          className="nameInput form-control"
          defaultValue=""
          onChange={[Function]}
          placeholder="Player 1 Name"
        />
        <input
          className="nameInput form-control"
          defaultValue=""
          onChange={[Function]}
          placeholder="Player 2 Name"
        />
      </div>
    </div>
    <div
      className="SizeSelect"
    >
      <div
        className="SettingsHeading"
      >
        <div
          className="ButtonCaption"
        >
          Board Size
        </div>
        <hr
          className="SettingsLine"
        />
      </div>
      <div
        className="SettingsValueRow"
      >
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          6
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          7
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          8
        </div>
      </div>
    </div>
  </div>
  <div
    className="SettingsButtonRow"
  >
    <div
      className="BackButtonYellow"
      onClick={[Function]}
    >
      <div
        className="ButtonCaption"
      >
        Back
      </div>
    </div>
    <div
      className="BackButtonGreen"
      onClick={[Function]}
    >
      <div
        className="ButtonCaption"
      >
        Start
      </div>
    </div>
  </div>
</div>
`;

exports[`GameSettings, renders correctly 1`] = `
<div
  className="MenuBoard"
>
  <div
    className="MenuTitle"
  >
    <div
      className="MenuTitleWhite"
    >
      Game
    </div>
    <div
      className="MenuTitleRed"
    >
      Settings
    </div>
  </div>
  <div
    className="SettingsRow"
  >
    <div
      className="NameSelect"
    >
      <div
        className="SettingsHeading"
      >
        <div
          className="ButtonCaption"
        >
          Enter Names
        </div>
        <hr
          className="SettingsLine"
        />
      </div>
      <div
        className="goOnlineButton"
        onClick={[Function]}
      >
        <div
          className="ButtonCaption"
        >
          Switch to Online Mode
        </div>
      </div>
      <div
        className="nameZone"
      >
        <input
          className="nameInput form-control"
          defaultValue=""
          onChange={[Function]}
          placeholder="Player 1 Name"
        />
        <input
          className="nameInput form-control"
          defaultValue=""
          onChange={[Function]}
          placeholder="Player 2 Name"
        />
      </div>
    </div>
    <div
      className="SizeSelect"
    >
      <div
        className="SettingsHeading"
      >
        <div
          className="ButtonCaption"
        >
          Board Size
        </div>
        <hr
          className="SettingsLine"
        />
      </div>
      <div
        className="SettingsValueRow"
      >
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          6
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          7
        </div>
        <div
          className="UnselectedValue"
          onClick={[Function]}
        >
          8
        </div>
      </div>
    </div>
  </div>
  <div
    className="SettingsButtonRow"
  >
    <div
      className="BackButtonYellow"
      onClick={[Function]}
    >
      <div
        className="ButtonCaption"
      >
        Back
      </div>
    </div>
    <div
      className="BackButtonGreen"
      onClick={[Function]}
    >
      <div
        className="ButtonCaption"
      >
        Start
      </div>
    </div>
  </div>
</div>
`;
